File: src/main/java/xf/xflp/base/container/constraints/BearingWeightQueue.java
package xf.xflp.base.container.constraints;

import xf.xflp.base.container.ZItemGraph;
import xf.xflp.base.item.Item;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Copyright (c) 2012-2022 Holger Schneider
 * All rights reserved.
 *
 * This source code is licensed under the MIT License (MIT) found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @author hschneid
 *
 * A queue of stacked items, where each item is returned next, when all upper items
 * are processed.
 */
public class BearingWeightQueue {

    private final boolean[] isProcessed;
    private final Map<Integer, int[]> upperItems = new HashMap<>();

    public BearingWeightQueue(int nbrOfItems) {
        this.isProcessed = new boolean[nbrOfItems];
    }

    /**
     * Adds an item to queue with all their upper items.
     *
     * The given item means as processed.
     */
    public void add(Item newItem, ZItemGraph graph) {
        List<Item> uppers = graph.getItemsAbove(newItem);
        int[] upperIdx = new int[uppers.size()];
        for (int i = upperIdx.length - 1; i >= 0; i--) {
            Item upperItem = uppers.get(i);
            upperIdx[i] = upperItem.index;

            // If unprocessed upper items were found,
            // add them to queue instead of this item
            if(!isProcessed[upperItem.index]) {
                add(upperItem, graph);
                return;
            }
        }

        upperItems.put(newItem.index, upperIdx);
    }

    /**
     * Get next item, where all upper items are processed.
     */
    public int getNext() {
        for (Integer idx : upperItems.keySet()) {
            if(areAllProcessed(idx)) {
                upperItems.remove(idx);
                return idx;
            }
        }
        return -1;
    }

    public boolean hasMore() {
        return upperItems.size() > 0;
    }

    private boolean areAllProcessed(int itemIdx) {
        int[] upperIdx = upperItems.get(itemIdx);
        boolean isOK = true;
        for (int i = upperIdx.length - 1; i >= 0; i--) {
            isOK &= isProcessed[upperIdx[i]];
        }

        return isOK;
    }

    public void setProcessed(int index) {
        isProcessed[index] = true;
    }
}

File: src/xflp/base/container/constraints/bearing_weight_queue.py
"""
Queue for processing items in bearing weight order.

Copyright (c) 2012-2025 Holger Schneider
Licensed under the MIT License
"""

from typing import Dict, List, TYPE_CHECKING

if TYPE_CHECKING:
    from ...item.item import Item
    from ..z_item_graph import ZItemGraph


class BearingWeightQueue:
    """
    A queue of stacked items where each item is returned when all upper items
    are processed.
    
    This ensures bearing weights are calculated top-down.
    """
    
    def __init__(self, nbr_of_items: int):
        """
        Initialize the queue.
        
        Args:
            nbr_of_items: Maximum number of items
        """
        self._is_processed: List[bool] = [False] * nbr_of_items
        self._upper_items: Dict[int, List[int]] = {}
    
    def add(self, new_item: 'Item', graph: 'ZItemGraph') -> None:
        """
        Add an item to queue with all their upper items.
        
        The given item is marked for processing.
        
        Args:
            new_item: The item to add
            graph: The Z-item graph
        """
        uppers = graph.get_items_above(new_item)
        upper_idx = [upper_item.index for upper_item in uppers]
        
        # If unprocessed upper items were found, add them instead
        for upper_item in uppers:
            if not self._is_processed[upper_item.index]:
                self.add(upper_item, graph)
                return
        
        self._upper_items[new_item.index] = upper_idx
    
    def get_next(self) -> int:
        """
        Get next item where all upper items are processed.
        
        Returns:
            The item index, or -1 if none available
        """
        for idx in list(self._upper_items.keys()):
            if self._are_all_processed(idx):
                del self._upper_items[idx]
                return idx
        return -1
    
    def has_more(self) -> bool:
        """
        Check if there are more items to process.
        
        Returns:
            True if items remain
        """
        return len(self._upper_items) > 0
    
    def _are_all_processed(self, item_idx: int) -> bool:
        """
        Check if all upper items of an item are processed.
        
        Args:
            item_idx: The item index
            
        Returns:
            True if all uppers are processed
        """
        upper_idx = self._upper_items.get(item_idx, [])
        return all(self._is_processed[idx] for idx in upper_idx)
    
    def set_processed(self, index: int) -> None:
        """
        Mark an item as processed.
        
        Args:
            index: The item index
        """
        if index < len(self._is_processed):
            self._is_processed[index] = True
